llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Enable curriculum learning by resuming with a different dataset.
+
+This callback is currently experimental. The API may change in the future.
+"""
+
+import logging
+from typing import Any, Dict, Iterable, Union
+
+from composer.core import Callback, State
+from composer.loggers import Logger
+from streaming import StreamingDataset
+from torch.utils.data import DataLoader
+
+log = logging.getLogger(__name__)
+
+
+class CurriculumLearning(Callback):
+    """Starts an epoch with a different dataset when resuming from a checkpoint.
+
+    This callback is currently experimental. The API may change in the future.
+
+    Args:
+        dataset_index (int): The index of the dataset currently being used.
+        current_dataset_config (Dict): The configuration of the dataset currently
+            being used.
+        dataset (StreamingDataset): The dataset currently being used.
```

Comment:
Currently you pass in dataloader and not dataset

Should this be a dataset instead?

Reply:
Oh I meant for this to be dataloader. I'd rather check if the dataset is a StreamingDataset in the callback rather than in build_callback or in trainer.py because it's particular to this callback alone, and keeps the existing code cleaner

Reply:
Well, because the dataloader is `Union[DataLoader, Iterable]` it won't always have the `dataset` attribute. So I'd rather just do the dataloader and dataset checks both inside `init` rather than separately

Reply:
with the `BEFORE_LOAD` event, still doing type checks inside there.